File: server/app/services/filename.py
"""Extraction employeur + intitulé de poste depuis une fiche + nommage fichier."""

from __future__ import annotations

import re
import unicodedata
LABEL_RE = re.compile(
    r"(?:^|\n)\s*(?:entreprise|soci[eé]t[eé]|employeur|company|client|organisme|structure|employer|groupe|cabinet|agence)\s*[:\-–—]\s*([^\n|,;]{2,80})",
    re.IGNORECASE,
)

LABEL_INLINE_RE = re.compile(
    r"(?:^|\n)\s*(?:entreprise|soci[eé]t[eé]|employeur|company|groupe)\s+"
    r"([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{1,40}"
    r"(?:\s+[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{1,30}){0,3})\b"
)

ABOUT_KW_RE = re.compile(
    r"(?:rejoignez|rejoindre|rejoins|au sein (?:de|d['’])|chez|pour le compte (?:de|d['’])|"
    r"au sein d['’]?une?|"
    r"[aà]\s*propos\s+de)\s+",
    re.IGNORECASE,
)

COMPANY_NAME_RE = re.compile(
    r"^([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{0,40}"
    r"(?:\s+(?:et|&|de|du|des|la|le)?\s*[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{1,30}){0,4})"
)

RECRUTE_RE = re.compile(
    r"(?:^|\n)\s*([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,40}"
    r"(?:\s+[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,30}){0,3})\s+"
    r"(?:recrute|recherche|embauche)",
    re.IGNORECASE,
)

DASH_LINE_RE = re.compile(
    r"(?:^|\n)\s*([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,40}"
    r"(?:\s+[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,30}){0,3})\s*[\-–—]\s*"
    r"(?:CDI|CDD|Stage|Alternance|H\/F|F\/H)",
    re.IGNORECASE,
)

JOB_TITLE_LABEL_RE = re.compile(
    r"(?:^|\n)\s*(?:intitul[eé]\s*(?:du\s*)?poste|titre\s*(?:du\s*)?poste|poste|"
    r"job\s*title|position|r[oô]le)\s*[:\-–—]\s*([^\n]{3,100})",
    re.IGNORECASE,
)

JOB_TITLE_HF_RE = re.compile(
    r"(?:^|\n)\s*([^\n]{5,90}?)\s*(?:\((?:H\/F|F\/H|HF|M\/F)\)|(?:H\/F|F\/H))\s*(?:\n|$)",
    re.IGNORECASE,
)

JOB_TITLE_INLINE_RE = re.compile(
    r"(?:nous\s+(?:recherchons|cherchons)|recherche(?:ons)?)\s+(?:un[e]?\s+)?"
    r"([^\n.]{5,80}?)(?:\s+(?:pour|afin|qui)\b|[.!]|$)",
    re.IGNORECASE,
)

NOISE = frozenset(
    {
        "cdi",
        "cdd",
        "temps",
        "plein",
        "paris",
        "france",
        "remote",
        "hybride",
        "junior",
        "senior",
        "h/f",
        "hf",
        "f/h",
        "nous",
        "notre",
        "nos",
        "votre",
        "vos",
        "une",
        "un",
        "le",
        "la",
        "les",
        "des",
        "du",
        "poste",
        "offre",
        "emploi",
        "equipe",
        "équipes",
        "equipes",
        "team",
        "teams",
    }
)


def _strip_accents(text: str) -> str:
    nfd = unicodedata.normalize("NFD", text)
    return "".join(c for c in nfd if unicodedata.category(c) != "Mn")


def slug_filename(name: str | None, *, max_len: int = 48) -> str:
    slug = _strip_accents(str(name or ""))
    slug = slug.lower()
    slug = re.sub(r"[^a-z0-9]+", "-", slug)
    slug = re.sub(r"^-|-$", "", slug)
    slug = slug[:max_len]
    slug = re.sub(r"-$", "", slug)
    return slug or ""


def extract_company(job_text: str | None, llm_company: str | None = None) -> str | None:
    """Récupère l'employeur : LLM d'abord, puis heuristiques multiples."""
    from_llm = _clean_company(llm_company)
    if from_llm:
        return from_llm

    text = str(job_text or "")
    candidates: list[str] = []

    labeled = LABEL_RE.search(text)
    if labeled:
        candidates.append(labeled.group(1))

    labeled_inline = LABEL_INLINE_RE.search(text)
    if labeled_inline:
        candidates.append(labeled_inline.group(1))

    recrute = RECRUTE_RE.search(text)
    if recrute:
        candidates.append(recrute.group(1))

    dash = DASH_LINE_RE.search(text)
    if dash:
        candidates.append(dash.group(1))

    about = _extract_about_company(text)
    if about:
        candidates.append(about)

    mail = re.search(r"@([a-z0-9-]+)\.(?:com|fr|io|co|net|org)\b", text, re.IGNORECASE)
    if mail and not re.search(r"gmail|outlook|hotmail|yahoo|proton", mail.group(1), re.IGNORECASE):
        candidates.append(_title_case_company(mail.group(1).replace("-", " ")))

    url = re.search(
        r"(?:https?://)?(?:www\.)?([a-z0-9-]+)\.(?:com|fr|io|co|net|org)\b",
        text,
        re.IGNORECASE,
    )
    if url and not re.search(
        r"linkedin|welcome|indeed|hellowork|google|github", url.group(1), re.IGNORECASE
    ):
        candidates.append(_title_case_company(url.group(1).replace("-", " ")))

    for raw in candidates:
        cleaned = _clean_company(raw)
        if cleaned:
            return cleaned
    return None


def extract_job_title(job_text: str | None) -> str | None:
    """Extrait l'intitulé de poste déclaré dans la fiche (si présent)."""
    text = str(job_text or "")

    labeled = JOB_TITLE_LABEL_RE.search(text)
    if labeled:
        title = _clean_job_title(labeled.group(1))
        if title:
            return title

    hf = JOB_TITLE_HF_RE.search(text)
    if hf:
        title = _clean_job_title(hf.group(1))
        if title:
            return title

    inline = JOB_TITLE_INLINE_RE.search(text)
    if inline and _looks_like_job_title(inline.group(1)):
        title = _clean_job_title(inline.group(1))
        if title:
            return title

    first = next(
        (
            line.strip()
            for line in text.split("\n")
            if 8 <= len(line.strip()) <= 80
        ),
        None,
    )
    if first and _looks_like_job_title(first) and not _looks_like_sentence(first):
        return _clean_job_title(first)
    return None


PERSON_SLUG = "lucas-schrever"


def build_job_export_filename(
    job_title: str | None,
    company: str | None,
    *,
    english: bool = False,
) -> str:
    """lucas-schrever-{poste}-{entreprise|defaut}-{fr|en}."""
    poste = slug_filename(job_title, max_len=48) or "cv"
    org = slug_filename(company, max_len=40) or ("default" if english else "defaut")
    lang = "en" if english else "fr"
    return f"{PERSON_SLUG}-{poste}-{org}-{lang}"


def _extract_about_company(text: str) -> str | None:
    for match in ABOUT_KW_RE.finditer(text):
        rest = text[match.end() :]
        name = COMPANY_NAME_RE.match(rest)
        if not name:
            continue
        cleaned = _clean_company(name.group(1))
        if cleaned:
            return cleaned
    return None


def _clean_company(raw: str | None) -> str | None:
    t = str(raw or "")
    t = re.sub(r"\s+", " ", t)
    t = re.sub(r"[|•·].*$", "", t)
    t = re.sub(r"\s*[-–—]\s*(cdi|cdd|stage|alternance|h/f|f/h).*$", "", t, flags=re.IGNORECASE)
    t = re.sub(r"\s*\([^)]*\)\s*$", "", t)
    t = re.sub(r"[.,;:!?]+$", "", t)
    t = t.strip()
    if not t or len(t) < 2 or len(t) > 80:
        return None

    if re.match(r"^(une?|des?|du|la|le|les|nos?|notre|votre|vos|cette?|cet)\b", t, re.IGNORECASE):
        return None
    if re.search(r"\b(équipe|equipes|équipes|teams?)\b", t, re.IGNORECASE) and not re.search(
        r"\b(SAS|SA|SARL|Inc|Ltd)\b", t, re.IGNORECASE
    ):
        return None

    norm = _strip_accents(t).lower()
    if norm in NOISE:
        return None
    if re.match(r"^(poste|offre|mission|profil|description|intitule|titre)$", t, re.IGNORECASE):
        return None
    if (
        re.match(
            r"^(developpeur|developer|ingenieur|ingénieur|chef|manager|expert|product|software)",
            t,
            re.IGNORECASE,
        )
        and len(t.split()) <= 3
    ):
        return None
    return t[:80]


def _clean_job_title(raw: str | None) -> str | None:
    t = str(raw or "")
    t = re.sub(r"\s+", " ", t)
    t = re.sub(r"\s*\((?:H/F|F/H|HF|M/F)\)\s*", " ", t, flags=re.IGNORECASE)
    t = re.sub(r"\b(?:H/F|F/H|HF|M/F)\b", " ", t, flags=re.IGNORECASE)
    t = re.sub(r"\s*[-–—]\s*(CDI|CDD|Stage|Alternance).*$", "", t, flags=re.IGNORECASE)
    t = re.sub(r"[#*_`]", "", t)
    t = re.sub(r"\s+", " ", t).strip()
    if not t or len(t) < 3 or len(t) > 90:
        return None
    if re.match(
        r"^(entreprise|societe|société|company|missions?|profil|description)",
        t,
        re.IGNORECASE,
    ):
        return None
    return t[:90]


def _looks_like_job_title(line: str) -> bool:
    return bool(
        re.search(
            r"developpeur|développeur|developer|engineer|ingénieur|chef de projet|"
            r"product owner|manager|consultant|expert|fullstack|full-stack|frontend|backend",
            line,
            re.IGNORECASE,
        )
    )


def _looks_like_sentence(line: str) -> bool:
    return bool(re.match(r"^(chez|nous|rejoignez|vous|notre|dans|pour|avec)\b", line, re.IGNORECASE)) or bool(
        re.search(r"[,.]", line)
    )


def _title_case_company(slug: str) -> str:
    return " ".join(
        w[0].upper() + w[1:].lower() if w else w
        for w in str(slug).split()
        if w
    )

File: server/src/services/filename.js
/**
 * Extraction employeur + intitulé de poste depuis une fiche + nommage fichier.
 */

const LABEL_RE =
  /(?:^|\n)\s*(?:entreprise|soci[eé]t[eé]|employeur|company|client|organisme|structure|employer|groupe|cabinet|agence)\s*[:\-–—]\s*([^\n|,;]{2,80})/i;

/** "Entreprise Acme", "Société Acme Digital" en début de ligne */
const LABEL_INLINE_RE =
  /(?:^|\n)\s*(?:entreprise|soci[eé]t[eé]|employeur|company|groupe)\s+([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{1,40}(?:\s+[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{1,30}){0,3})\b/;

const ABOUT_KW_RE =
  /(?:rejoignez|rejoindre|au sein (?:de|d['’])|chez|pour le compte (?:de|d['’])|au sein d['’]?une?)\s+/gi;

const COMPANY_NAME_RE =
  /^([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{0,40}(?:\s+(?:et|&|de|du|des|la|le)?\s*[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ0-9][\wÀ-ÿ&.'’-]{1,30}){0,4})/;

/** "Acme Digital recrute", "ACME - CDI", "Offre Acme Digital" */
const RECRUTE_RE =
  /(?:^|\n)\s*([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,40}(?:\s+[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,30}){0,3})\s+(?:recrute|recherche|embauche)/i;

const DASH_LINE_RE =
  /(?:^|\n)\s*([A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,40}(?:\s+[A-ZÀÂÄÉÈÊËÏÎÔÙÛÜÇ][\wÀ-ÿ&.'’-]{1,30}){0,3})\s*[\-–—]\s*(?:CDI|CDD|Stage|Alternance|H\/F|F\/H)/i;

const JOB_TITLE_LABEL_RE =
  /(?:^|\n)\s*(?:intitul[eé]\s*(?:du\s*)?poste|titre\s*(?:du\s*)?poste|poste|job\s*title|position|r[oô]le)\s*[:\-–—]\s*([^\n]{3,100})/i;

const JOB_TITLE_HF_RE =
  /(?:^|\n)\s*([^\n]{5,90}?)\s*\((?:H\/F|F\/H|HF|M\/F)\)/i;

const JOB_TITLE_INLINE_RE =
  /(?:nous\s+(?:recherchons|cherchons)|recherche(?:ons)?)\s+(?:un[e]?\s+)?([^\n.]{5,80}?)(?:\s+(?:pour|afin|qui)\b|[.!]|$)/i;

const NOISE = new Set([
  "cdi",
  "cdd",
  "temps",
  "plein",
  "paris",
  "france",
  "remote",
  "hybride",
  "junior",
  "senior",
  "h/f",
  "hf",
  "f/h",
  "nous",
  "notre",
  "nos",
  "votre",
  "vos",
  "une",
  "un",
  "le",
  "la",
  "les",
  "des",
  "du",
  "poste",
  "offre",
  "emploi",
  "equipe",
  "équipes",
  "equipes",
  "team",
  "teams",
]);

export function slugFilename(name, { maxLen = 48 } = {}) {
  return (
    String(name || "")
      .normalize("NFD")
      .replace(/\p{M}/gu, "")
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "")
      .slice(0, maxLen)
      .replace(/-$/, "") || ""
  );
}

/**
 * Récupère l'employeur : LLM d'abord, puis heuristiques multiples.
 * @returns {string|null}
 */
export function extractCompany(jobText, llmCompany = null) {
  const fromLlm = cleanCompany(llmCompany);
  if (fromLlm) return fromLlm;

  const text = String(jobText || "");
  // Ordre de priorité (labels / recrute avant "chez/rejoindre" trop ambigus)
  const candidates = [];

  const labeled = text.match(LABEL_RE);
  if (labeled) candidates.push(labeled[1]);

  const labeledInline = text.match(LABEL_INLINE_RE);
  if (labeledInline) candidates.push(labeledInline[1]);

  const recrute = text.match(RECRUTE_RE);
  if (recrute) candidates.push(recrute[1]);

  const dash = text.match(DASH_LINE_RE);
  if (dash) candidates.push(dash[1]);

  const about = extractAboutCompany(text);
  if (about) candidates.push(about);

  // Domaine email / URL type careers.acme.com
  const mail = text.match(/@([a-z0-9-]+)\.(?:com|fr|io|co|net|org)\b/i);
  if (mail && !/gmail|outlook|hotmail|yahoo|proton/i.test(mail[1])) {
    candidates.push(titleCaseCompany(mail[1].replace(/-/g, " ")));
  }
  const url = text.match(
    /(?:https?:\/\/)?(?:www\.)?([a-z0-9-]+)\.(?:com|fr|io|co|net|org)\b/i,
  );
  if (url && !/linkedin|welcome|indeed|hellowork|google|github/i.test(url[1])) {
    candidates.push(titleCaseCompany(url[1].replace(/-/g, " ")));
  }

  for (const raw of candidates) {
    const c = cleanCompany(raw);
    if (c) return c;
  }
  return null;
}

/**
 * Extrait l'intitulé de poste déclaré dans la fiche (si présent).
 * @returns {string|null}
 */
export function extractJobTitle(jobText) {
  const text = String(jobText || "");
  const labeled = text.match(JOB_TITLE_LABEL_RE);
  if (labeled) {
    const t = cleanJobTitle(labeled[1]);
    if (t) return t;
  }
  const hf = text.match(JOB_TITLE_HF_RE);
  if (hf) {
    const t = cleanJobTitle(hf[1]);
    if (t) return t;
  }
  const inline = text.match(JOB_TITLE_INLINE_RE);
  if (inline && looksLikeJobTitle(inline[1])) {
    const t = cleanJobTitle(inline[1]);
    if (t) return t;
  }
  // Première ligne courte qui ressemble à un titre
  const first = text
    .split(/\n/)
    .map((l) => l.trim())
    .find((l) => l.length >= 8 && l.length <= 80);
  if (first && looksLikeJobTitle(first) && !looksLikeSentence(first)) {
    return cleanJobTitle(first);
  }
  return null;
}

function extractAboutCompany(text) {
  ABOUT_KW_RE.lastIndex = 0;
  let m;
  while ((m = ABOUT_KW_RE.exec(text)) !== null) {
    const rest = text.slice(m.index + m[0].length);
    // Préférer un nom propre (capitale)
    let name = rest.match(COMPANY_NAME_RE);
    if (!name) continue;
    const c = cleanCompany(name[1]);
    if (c) return c;
  }
  return null;
}

function cleanCompany(raw) {
  let t = String(raw || "")
    .replace(/\s+/g, " ")
    .replace(/[|•·].*$/, "")
    .replace(/\s*[-–—]\s*(cdi|cdd|stage|alternance|h\/f|f\/h).*$/i, "")
    .replace(/\s*\([^)]*\)\s*$/g, "")
    .replace(/[.,;:!?]+$/g, "")
    .trim();
  if (!t || t.length < 2 || t.length > 80) return null;

  // Refuse fragments de phrase / articles
  if (
    /^(une?|des?|du|la|le|les|nos?|notre|votre|vos|cette?|cet)\b/i.test(t)
  ) {
    return null;
  }
  if (/\b(équipe|equipes|équipes|teams?)\b/i.test(t) && !/\b(SAS|SA|SARL|Inc|Ltd)\b/i.test(t)) {
    return null;
  }

  const norm = t
    .normalize("NFD")
    .replace(/\p{M}/gu, "")
    .toLowerCase();
  if (NOISE.has(norm)) return null;
  if (/^(poste|offre|mission|profil|description|intitule|titre)$/i.test(t)) return null;
  if (
    /^(developpeur|developer|ingenieur|ingénieur|chef|manager|expert|product|software)/i.test(
      t,
    ) &&
    t.split(/\s+/).length <= 3
  ) {
    // Évite de prendre un intitulé de poste pour une entreprise
    return null;
  }
  return t.slice(0, 80);
}

function cleanJobTitle(raw) {
  let t = String(raw || "")
    .replace(/\s+/g, " ")
    .replace(/\s*\((?:H\/F|F\/H|HF|M\/F)\)\s*/gi, " ")
    .replace(/\b(?:H\/F|F\/H|HF|M\/F)\b/gi, " ")
    .replace(/\s*[-–—]\s*(CDI|CDD|Stage|Alternance).*$/i, "")
    .replace(/[#*_`]/g, "")
    .replace(/\s+/g, " ")
    .trim();
  if (!t || t.length < 3 || t.length > 90) return null;
  if (/^(entreprise|societe|société|company|missions?|profil|description)/i.test(t)) return null;
  return t.slice(0, 90);
}

function looksLikeJobTitle(line) {
  return /developpeur|développeur|developer|engineer|ingénieur|chef de projet|product owner|manager|consultant|expert|fullstack|full-stack|frontend|backend/i.test(
    line,
  );
}

function looksLikeSentence(line) {
  return /^(chez|nous|rejoignez|vous|notre|dans|pour|avec)\b/i.test(line) || /[,.]/.test(line);
}

function titleCaseCompany(slug) {
  return String(slug)
    .split(/\s+/)
    .filter(Boolean)
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase())
    .join(" ");
}

/**
 * Nom de fichier : {titre}-{entreprise}
 */
export function buildJobExportFilename(profileName, company) {
  const base = slugFilename(profileName, { maxLen: 48 }) || "cv";
  const org = slugFilename(company, { maxLen: 40 });
  return org ? `${base}-${org}` : base;
}
